src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,229 @@
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);
+                            }
+
+                            bulkUpdateGraph.AddRelatedDescriptor(topLevelDescriptor, entityDescriptor);
+                            this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
+                        }
+                    }
+                    else if (descriptor is LinkDescriptor linkDescriptor)
+                    {
+                        if (linkDescriptor.Source.Equals(topLevelDescriptor.Entity) && !bulkUpdateGraph.Contains(linkDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (linkDescriptor.State == EntityStates.Deleted || linkDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);
+                            }
+
+                            EntityDescriptor targetDescriptor = this.RequestInfo.Context.GetEntityDescriptor(linkDescriptor.Target);
+
+                            if (targetDescriptor != null && (targetDescriptor.State == EntityStates.Deleted || targetDescriptor.State == EntityStates.Modified))
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);
+                            }
+
+                            bulkUpdateGraph.AddRelatedDescriptor(topLevelDescriptor, linkDescriptor);
+                            this.BuildDescriptorGraph(descriptors, false, linkDescriptor.Target);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates an ODataRequestMessage for a deep insert request.
+        /// </summary>
+        /// <returns>Returns an instance of the ODataRequestMessage for the deep insert request.</returns>
+        private ODataRequestMessageWrapper CreateDeepInsertRequest(BulkUpdateGraph bulkUpdateGraph)

```

Comment:
@KenitoInc I don't think any of your tests hit this method. The lines starting with `headers.SetHeader` both throw the following exception:
```csharp
The value 'V401' of the OData-Version HTTP header is invalid. Only '4.0' and '4.01' are supported as values for the OData-Version header.
```
Without pulling your branch and testing out the feature I wouldn't have caught this... If you had tests covering the changes in this PR, you would have caught this.

Reply:
This should be covered in E2E tests

Reply:
@KenitoInc If that is the case, leave it for PR where you'll add the tests